crates/red_knot_python_semantic/src/types/generics.rs

```diff
@@ -146,3 +156,114 @@ impl<'db> Specialization<'db> {
             .map(|(_, ty)| *ty)
     }
 }
+
+/// Performs type inference between parameter annotations and argument types, producing a
+/// specialization of a generic function.
+pub(crate) struct SpecializationBuilder<'db> {
+    db: &'db dyn Db,
+    generic_context: GenericContext<'db>,
+    types: FxHashMap<TypeVarInstance<'db>, UnionBuilder<'db>>,
+}
+
+impl<'db> SpecializationBuilder<'db> {
+    pub(crate) fn new(db: &'db dyn Db, generic_context: GenericContext<'db>) -> Self {
+        Self {
+            db,
+            generic_context,
+            types: FxHashMap::default(),
+        }
+    }
+
+    pub(crate) fn build(mut self) -> Specialization<'db> {
+        let types = self
+            .generic_context
+            .variables(self.db)
+            .iter()
+            .map(|variable| {
+                self.types
+                    .remove(variable)
+                    .map(UnionBuilder::build)
+                    .unwrap_or(variable.default_ty(self.db).unwrap_or(Type::unknown()))
+            })
+            .collect();
+        Specialization::new(self.db, self.generic_context, types)
+    }
+
+    fn add_type_mapping(&mut self, typevar: TypeVarInstance<'db>, ty: Type<'db>) {
+        let builder = self
+            .types
+            .entry(typevar)
+            .or_insert_with(|| UnionBuilder::new(self.db));
+        builder.add_in_place(ty);
```

Comment:
Here is where (as discussed above) we are unioning together the mappings for a typevar that appears in multiple parameters